Actor checkpointing for distributed actor handles

stephanie-wang opened the pull request:
## What do these changes do?

This replaces the current actor checkpointing mechanism and extends it to support distributed actor handles.

After executing a specified interval of tasks, the actor will execute a method to save the checkpoint. The checkpoint state includes the local scheduler's current task frontier for the actor, which is a mapping from handle ID to task counter and object dependency.

When an actor executes its first task (`__init__`), it checks first for a valid checkpoint. If one exists, the actor initializes its state from the saved checkpoint. It also resets the local scheduler's current task frontier. The task counters allow the local scheduler to filter out duplicate tasks submitted during reconstruction. The object dependencies allow the local scheduler to fulfill execution dependencies for tasks that execute after the checkpoint.

## Related issue number

None

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/3512/
Test PASSed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/3521/
Test PASSed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/3528/
Test PASSed.

AmplabJenkins commented:
Build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/3541/
Test FAILed.

robertnishihara commented on python/ray/actor.py:
```diff
@@ -93,73 +100,123 @@ def get_actor_checkpoint(worker, actor_id):
+
+        # If this is the first task to execute on the actor, try to resume from
+        # a checkpoint.
+        if actor_imported and worker.actor_task_counter == 1:
```
Will this be something that we can do during the "actor creation task" once we make actor creation a task?

stephanie-wang replied:
Hmm that depends on what the actor creation task does...if it imports the actor class, then yes.

robertnishihara commented on src/local_scheduler/local_scheduler_algorithm.cc:
```diff
@@ -1580,3 +1619,74 @@ void print_worker_info(const char *message,
+void set_actor_task_counters(
+    SchedulingAlgorithmState *algorithm_state,
+    ActorID actor_id,
+    std::unordered_map<ActorID, int64_t, UniqueIDHasher> task_counters) {
```
const ref for `actor_id` and `task_counters`?

robertnishihara commented on src/local_scheduler/local_scheduler_algorithm.cc:
```diff
@@ -1580,3 +1619,74 @@ void print_worker_info(const char *message,
+
+std::unordered_map<ActorID, int64_t, UniqueIDHasher> get_actor_task_counters(
+    SchedulingAlgorithmState *algorithm_state,
+    ActorID actor_id) {
```
const ref

robertnishihara commented on src/local_scheduler/local_scheduler_algorithm.cc:
```diff
@@ -1580,3 +1619,74 @@ void print_worker_info(const char *message,
+
+std::unordered_map<ActorID, ObjectID, UniqueIDHasher> get_actor_frontier(
+    SchedulingAlgorithmState *algorithm_state,
+    ActorID actor_id) {
```
const ref

robertnishihara commented on src/local_scheduler/local_scheduler_algorithm.cc:
```diff
@@ -1580,3 +1619,74 @@ void print_worker_info(const char *message,
+  entry.task_counters = task_counters;
+
+  /* Filter out duplicate tasks for the actor that are runnable and that were
+   * submitted earlier than the new task counter. */
```
When can tasks be filtered out like this? Can this only happen during reconstruction if a task is resubmitted and we end up with two copies? Might be worth explaining how this situation can arise in a comment.

robertnishihara commented on src/local_scheduler/local_scheduler_algorithm.h:
```diff
@@ -307,6 +304,65 @@ int reconstruct_object_timeout_handler(event_loop *loop,
+ */
+void set_actor_task_counters(
+    SchedulingAlgorithmState *algorithm_state,
+    ActorID actor_id,
```
const ref, same comment applies in a bunch of places

concretevitamin replied:
+1; const -> happiness

robertnishihara commented on src/local_scheduler/local_scheduler_client.cc:
```diff
@@ -102,13 +102,8 @@ void local_scheduler_submit(LocalSchedulerConnection *conn,
-  write_message(conn->conn, MessageType_GetTask, fbb.GetSize(),
-                fbb.GetBufferPointer());
+                                   int64_t *task_size) {
+  write_message(conn->conn, MessageType_GetTask, 0, NULL);
```
Interesting, I didn't realize we could do that

robertnishihara commented on src/local_scheduler/local_scheduler_client.cc:
```diff
@@ -177,3 +172,34 @@ void local_scheduler_put_object(LocalSchedulerConnection *conn,
+  int64_t type;
+  int64_t reply_size;
+  uint8_t *reply;
+  read_message(conn->conn, &type, &reply_size, &reply);
```
may make sense to use `read_vector` or that may be about the same

robertnishihara commented on python/ray/actor.py:
```diff
@@ -802,97 +837,55 @@ def __ray_restore_from_checkpoint__(cls, pickled_checkpoint):
+
+            This task looks for a saved checkpoint and if found, restores the
+            state of the actor, the task frontier in the local scheduler, and
+            the checkpoint index (number of tasks executed so far).
```
Document `Returns:`

robertnishihara commented on python/ray/actor.py:
```diff
@@ -802,97 +837,55 @@ def __ray_restore_from_checkpoint__(cls, pickled_checkpoint):
+                    worker.actor_class.__ray_restore_from_checkpoint__(
+                        checkpoint))
+                # Set the number of tasks executed so far.
+                worker.actor_task_counter = int(checkpoint_index)
```
Instead of casting to `int` here, this cast should probably be done inside of `get_actor_checkpoint`

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/3547/
Test PASSed.

robertnishihara commented on python/ray/actor.py:
```diff
@@ -93,73 +100,123 @@ def get_actor_checkpoint(worker, actor_id):
             method defined on the actor class and should therefore take an
             instance of the actor as the first argument.
+        actor_imported (bool): Whether the actor has been imported.
+            Checkpointing operations will not be run if this is set to False.
```
I see, this bool is just to avoid loading checkpoints in the case where unpickling the actor class raises an exception, right?

stephanie-wang replied:
Yes, there was an error where the checkpointing methods couldn't be found if the actor hasn't been imported yet.

robertnishihara commented on python/ray/actor.py:
```diff
@@ -93,73 +100,123 @@ def get_actor_checkpoint(worker, actor_id):
+                save_and_log_checkpoint(worker, actor)
+
+        if method_error is not None:
+            raise method_error
```
does this preserve the exception backtrace when we reraise the exception here? I can give this a try to see

stephanie-wang replied:
Nice catch! It doesn't; we have to re-raise within the `except` block to preserve the backtrace.

stephanie-wang commented:
@robertnishihara I believe I addressed all your comments, except for adding `const ref` to the `ActorID` arguments. We don't currently do that anywhere else in the local scheduler for `UniqueID`s, and I also am not sure it's really necessary for `UniqueID`s anyway.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/3567/
Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/3572/
Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/3588/
Test PASSed.

concretevitamin commented on src/local_scheduler/test/local_scheduler_tests.cc:
```diff
@@ -210,12 +210,12 @@ TEST object_reconstruction_test(void) {
     int64_t task_assigned_size;
     local_scheduler_submit(worker, execution_spec);
     TaskSpec *task_assigned =
-        local_scheduler_get_task(worker, &task_assigned_size, true);
```
Nice to have it removed, looking at this I had no idea what true means.  For literal arguments, especially booleans, I usually do
```
local_scheduler_get_task(worker, &task_assigned_size, /*param_name=*/true);
```

concretevitamin commented on src/local_scheduler/local_scheduler_client.h:
```diff
@@ -148,4 +145,25 @@ void local_scheduler_put_object(LocalSchedulerConnection *conn,
+ *
+ * @param conn The connection information.
+ * @param actor_id The ID of the actor whose frontier is returned.
+ * @return An ActorFrontier flatbuffer.
```
Return type != what's documented

concretevitamin commented on src/local_scheduler/local_scheduler_client.cc:
```diff
@@ -177,3 +172,29 @@ void local_scheduler_put_object(LocalSchedulerConnection *conn,
+void local_scheduler_set_actor_frontier(LocalSchedulerConnection *conn,
+                                        const std::vector<uint8_t> &frontier) {
+  write_message(conn->conn, MessageType_ActorFrontier, frontier.size(),
+                (uint8_t *) frontier.data());
```
Nit: use C++ style cast
http://google.github.io/styleguide/cppguide.html#Casting

concretevitamin commented on src/local_scheduler/local_scheduler_client.cc:
```diff
@@ -177,3 +172,29 @@ void local_scheduler_put_object(LocalSchedulerConnection *conn,
                 fbb.GetBufferPointer());
 }
+
+// std::unordered_map<ActorID, std::pair<int64_t, ObjectID>, UniqueIDHasher>
```
why the comment?

concretevitamin commented on src/local_scheduler/local_scheduler_algorithm.h:
```diff
@@ -307,6 +314,67 @@ int reconstruct_object_timeout_handler(event_loop *loop,
+        &task_counters);
+
+/**
+ * Get the actor's frontier of task dependencies.
```
Since "frontier" is a new concept introduced in this PR -- and primarily abstracted away by these two functions? -- maybe add a clear explanation/intuition on what it is before these two functions?

pcmoritz replied:
Yes, that a good idea!

concretevitamin commented on src/local_scheduler/format/local_scheduler.fbs:
```diff
@@ -97,3 +91,24 @@ table PutObject {
   object_id: string;
 }
+
+table ActorFrontier {
```
Briefly document what a Frontier means (intuitively) and what it's used for?

robertnishihara commented:
Posting here so I don't lose this script.

The following script hangs for me (pretty frequently) on MacOS, so this may explain some of the Travis errors.

```python
import ray
import time

ray.worker._init(start_ray_local=True, num_local_schedulers=2)

@ray.remote(checkpoint_interval=1)
class Counter(object):
    def __init__(self):
        pass

    def local_plasma(self):
        return ray.worker.global_worker.plasma_client.store_socket_name

    def method(self):
        time.sleep(1)


local_plasma = ray.worker.global_worker.plasma_client.store_socket_name

@ray.remote
def f(actor):
    ray.get(actor.method.remote())

actor = Counter.remote()
while ray.get(actor.local_plasma.remote()) == local_plasma:
    actor = Counter.remote()

time.sleep(5)

x_ids = [f.remote(actor) for _ in range(20)]

time.sleep(1)

process = ray.services.all_processes[ray.services.PROCESS_TYPE_PLASMA_STORE][1]
process.kill()
process.wait()

time.sleep(3)

print("Waiting at get!!")
ray.get(x_ids)
```

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/3614/
Test FAILed.

stephanie-wang commented:
@robertnishihara thanks! This fails because each handle only submits one task and so they share the same "frontier dependency". The local scheduler ends up trying to mark the same object as locally available twice. I fixed this in the most recent commit.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/3615/
Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/3616/
Test PASSed.

pcmoritz commented on src/local_scheduler/local_scheduler.cc:
```diff
@@ -1211,6 +1249,15 @@ void process_message(event_loop *loop,
+    ActorID actor_id = from_flatbuf(*message->actor_id());
+    handle_get_actor_frontier(state, worker, actor_id);
+  } break;
+  case MessageType_ActorFrontier: {
```
maybe rename this to MessageType_SetActorFrontierRequest?

pcmoritz commented on src/local_scheduler/local_scheduler.cc:
```diff
@@ -1066,6 +1054,59 @@ void handle_client_disconnect(LocalSchedulerState *state,
+      get_actor_task_counters(state->algorithm_state, actor_id);
+  auto frontier = get_actor_frontier(state->algorithm_state, actor_id);
+
+  std::vector<ActorID> handle_vector;
```
Should this be ActorHandleID?

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/3618/
Test PASSed.

concretevitamin commented:
LGTM from the language side

robertnishihara commented:
Leaving a note here for future reference that this PR probably introduces a **bug** (we've seen some hanging on Travis). However, it has been difficult to reproduce.

stephanie-wang commented:
Thanks, @robertnishihara. Just to clarify, these are the tests that we saw hanging, both on OSX:
`python test/actor_test.py ActorReconstruction.testCheckpointing`
`python test/actor_test.py ActorReconstruction.testLostCheckpoint`
